File: testdata/golden/api/generate.go
// Command generate renders the library API architecture template to PDF.
//
// Run from the repository root (also invoked by `make samples`):
//
//	go run ./testdata/golden/api
//
// Writes (overwriting if present):
//  1. output/architecture-diagram.pdf — sample viewer artifact
//
// It does not write testdata/golden/architecture-diagram.html or
// testdata/golden/api/architecture-diagram.pdf. The HTML corpus fixture is
// separate. fixture-56-architecture-diagram.html is a third, 20-page
// template. The only HTML this command reads is
// testdata/golden/api/architecture-diagram.html. Pass -output to send the
// PDF somewhere else; testdata/golden stays a source tree.
package main

import (
	"bytes"
	"context"
	"errors"
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	gowkhtmltopdf "github.com/chinmay-sawant/gowkhtmltopdf"
)

const (
	apiDirectory    = "testdata/golden/api"
	sampleDirectory = "output"
	inputName       = "architecture-diagram.html"
	outputName      = "architecture-diagram.pdf"
	pdfFileMode     = 0o600
	wantPages       = 5
)

var (
	errUnexpectedArguments = errors.New("unexpected command-line arguments")
	errUnexpectedPageCount = errors.New("unexpected generated page count")
	errTemplateNotFound    = errors.New("template not found")
	errTemplateDirectory   = errors.New("template path is a directory")
)

func main() {
	if err := run(os.Args[1:]); err != nil {
		fmt.Fprintln(os.Stderr, "generate:", err)
		os.Exit(1)
	}
}

func run(args []string) error { //nolint:cyclop,funlen // generator phases
	flags := flag.NewFlagSet("generate", flag.ContinueOnError)
	output := flags.String("output", "", "PDF output path (default: output/architecture-diagram.pdf)")

	if err := flags.Parse(args); err != nil {
		return fmt.Errorf("parse flags: %w", err)
	}

	if flags.NArg() != 0 {
		return fmt.Errorf("%w: %v", errUnexpectedArguments, flags.Args())
	}

	input, defaultOutput, _, err := resolveTemplatePaths()
	if err != nil {
		return err
	}

	if *output == "" {
		*output = defaultOutput
	}

	converter := gowkhtmltopdf.NewConverter()

	for _, setting := range []struct {
		name  string
		value string
	}{
		{name: "size.pagesize", value: "A4"},
		{name: "margin.top", value: "0"},
		{name: "margin.right", value: "0"},
		{name: "margin.bottom", value: "0"},
		{name: "margin.left", value: "0"},
		{name: "web.background", value: "true"},
		{name: "smartshrinking", value: "false"},
	} {
		if err := converter.Global().Set(setting.name, setting.value); err != nil {
			return fmt.Errorf("set global %q: %w", setting.name, err)
		}
	}

	if err := converter.Global().Set("enablelocalfileaccess", "true"); err != nil {
		return fmt.Errorf("enable local file access: %w", err)
	}

	page := gowkhtmltopdf.NewObjectSettings().SetPage(input)
	if err := page.Set("load.blocklocalfileaccess", "false"); err != nil {
		return fmt.Errorf("allow local template: %w", err)
	}

	converter.AddObject(page)

	if err := converter.Convert(context.Background()); err != nil {
		return fmt.Errorf("convert template: %w", err)
	}

	pdf := converter.Output()
	if got := pageCount(pdf); got != wantPages {
		return fmt.Errorf("%w: %s pages = %d, want %d", errUnexpectedPageCount, input, got, wantPages)
	}

	if err := writeFile(*output, pdf, pdfFileMode); err != nil {
		return fmt.Errorf("write PDF %s: %w", *output, err)
	}

	if _, err := fmt.Fprintf(os.Stdout, "generated %s (%d pages, %d bytes)\n", *output, wantPages, len(pdf)); err != nil {
		return fmt.Errorf("report generation: %w", err)
	}

	return nil
}

func pageCount(pdf []byte) int {
	return bytes.Count(pdf, []byte("/Type /Page\n"))
}

func writeFile(path string, data []byte, mode os.FileMode) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return fmt.Errorf("create parent directory: %w", err)
	}

	if err := os.WriteFile(path, data, mode); err != nil {
		return fmt.Errorf("write file: %w", err)
	}

	return nil
}

func uniquePaths(paths ...string) []string {
	out := make([]string, 0, len(paths))
	seen := make(map[string]struct{}, len(paths))

	for _, path := range paths {
		if path == "" {
			continue
		}

		absolute, err := filepath.Abs(path)
		if err != nil {
			// Fall back to the raw path; writeFile will surface any real error.
			absolute = path
		}

		if _, exists := seen[absolute]; exists {
			continue
		}

		seen[absolute] = struct{}{}
		out = append(out, absolute)
	}

	return out
}

func samePath(a, b string) bool {
	absA, errA := filepath.Abs(a)
	absB, errB := filepath.Abs(b)
	if errA != nil || errB != nil {
		return a == b
	}

	return absA == absB
}

// isAPITemplate reports whether path is the library-API source
// (testdata/golden/api/architecture-diagram.html), not the same-named
// corpus fixture at testdata/golden/architecture-diagram.html.
func isAPITemplate(path string) bool {
	return filepath.Base(filepath.Dir(path)) == "api" && filepath.Base(path) == inputName
}

// resolveTemplatePaths accepts invocation from the repository root, the api
// directory itself, or a compiled copy whose source directory is available.
// The returned input is absolute so the loader cannot reinterpret it as an
// HTTP host when the caller's working directory differs. repoRoot is the
// directory that contains testdata/ and output/.
func resolveTemplatePaths() (input, defaultOutput, repoRoot string, err error) {
	workingDir, err := os.Getwd()
	if err != nil {
		return "", "", "", fmt.Errorf("resolve working directory: %w", err)
	}

	_, sourceFile, _, sourceOK := runtime.Caller(0)
	sourceDir := ""

	if sourceOK {
		sourceDir = filepath.Dir(sourceFile)
	}

	candidates := []string{
		filepath.Join(workingDir, inputName),
		filepath.Join(workingDir, apiDirectory, inputName),
		filepath.Join(sourceDir, inputName),
	}
	checked := make([]string, 0, len(candidates))
	seen := make(map[string]struct{}, len(candidates))

	for _, candidate := range candidates {
		absolute, err := filepath.Abs(candidate)
		if err != nil {
			return "", "", "", fmt.Errorf("resolve template path %q: %w", candidate, err)
		}

		if _, exists := seen[absolute]; exists {
			continue
		}

		seen[absolute] = struct{}{}

		checked = append(checked, absolute)

		info, err := os.Stat(absolute)
		if errors.Is(err, os.ErrNotExist) {
			continue
		}

		if err != nil {
			return "", "", "", fmt.Errorf("inspect template %q: %w", absolute, err)
		}

		if info.IsDir() {
			return "", "", "", fmt.Errorf("%w: %s", errTemplateDirectory, absolute)
		}

		if !isAPITemplate(absolute) {
			// testdata/golden/architecture-diagram.html shares the basename
			// but is a corpus fixture, not this generator's source.
			continue
		}

		apiDir := filepath.Dir(absolute)
		// template lives at <repo>/testdata/golden/api/<file>
		root := filepath.Dir(filepath.Dir(filepath.Dir(apiDir)))

		return absolute, filepath.Join(root, sampleDirectory, outputName), root, nil
	}

	return "", "", "", fmt.Errorf("%w %q; checked %s", errTemplateNotFound, inputName, strings.Join(checked, ", "))
}

File: testdata/golden/python_api/generate.py
#!/usr/bin/env python3
"""Render the Python API architecture template to PDF.

This is the file-on-disk twin of ``testdata/golden/api/generate.go``: resolve
``architecture-diagram.html``, build a ``Document`` with the same option bag
(page size, zero margins, background, smart_shrinking, allow_local_files),
convert, assert page count, write bytes.

For the typical Python utility shape (HTML as in-memory bytes, no template
file), see ``generate_inline.py``.

Run from the repository root (also invoked by ``make python-api``):

    python3 testdata/golden/python_api/generate.py

Writes (overwriting if present):

  1. output/python/architecture-diagram.pdf

The v1 one-shot ABI accepts inline HTML only, so this reads the template
file and passes ``Content.from_html`` with a ``file://`` base URL (same
pattern as ``convert_file_to_pdf``). Go's generator can use ``File(path)``
directly; Python cannot until the handle-based ABI lands.
"""
import argparse
import os
import sys
from pathlib import Path

# Allow in-tree runs before pip install -e.
_REPO_CANDIDATES = [
    Path(__file__).resolve().parents[3],
    Path.cwd(),
]
for _root in _REPO_CANDIDATES:
    _src = _root / "bindings" / "python" / "src"
    if _src.is_dir():
        sys.path.insert(0, str(_src))
        break

from gowkhtmltopdf import Content, Document, Margin, Page  # noqa: E402

API_DIRECTORY = "testdata/golden/python_api"
SAMPLE_DIRECTORY = "output/python"
INPUT_NAME = "architecture-diagram.html"
OUTPUT_NAME = "architecture-diagram.pdf"
PDF_FILE_MODE = 0o600
WANT_PAGES = 5


class TemplateNotFoundError(FileNotFoundError):
    """Raised when architecture-diagram.html cannot be resolved."""


class UnexpectedPageCountError(RuntimeError):
    """Raised when the rendered PDF page count is not WANT_PAGES."""


def page_count(pdf):
    # type: (bytes) -> int
    return pdf.count(b"/Type /Page\n")


def is_python_api_template(path):
    # type: (Path) -> bool
    """True when path is testdata/golden/python_api/architecture-diagram.html."""
    return path.name == INPUT_NAME and path.parent.name == "python_api"


def unique_paths(*paths):
    # type: (str) -> list
    out = []
    seen = set()
    for path in paths:
        if not path:
            continue
        absolute = str(Path(path).resolve())
        if absolute in seen:
            continue
        seen.add(absolute)
        out.append(absolute)
    return out


def same_path(a, b):
    # type: (str, str) -> bool
    try:
        return Path(a).resolve() == Path(b).resolve()
    except OSError:
        return a == b


def resolve_template_paths(working_dir=None, source_file=None):
    # type: (...) -> tuple
    """Return (input, default_output, repo_root).

    Accepts invocation from the repository root, the python_api directory
    itself, or a compiled copy whose source directory is available.
    """
    cwd = working_dir if working_dir is not None else Path.cwd()
    source_dir = (
        source_file.parent
        if source_file is not None
        else Path(__file__).resolve().parent
    )

    candidates = [
        cwd / INPUT_NAME,
        cwd / API_DIRECTORY / INPUT_NAME,
        source_dir / INPUT_NAME,
    ]
    checked = []
    seen = set()

    for candidate in candidates:
        absolute = candidate.resolve()
        key = str(absolute)
        if key in seen:
            continue
        seen.add(key)
        checked.append(key)

        if not absolute.exists():
            continue
        if absolute.is_dir():
            raise IsADirectoryError(
                "template path is a directory: {0}".format(absolute)
            )
        if not is_python_api_template(absolute):
            # testdata/golden/architecture-diagram.html and api/ share the
            # basename but are not this generator's source.
            continue

        api_dir = absolute.parent
        # template lives at <repo>/testdata/golden/python_api/<file>
        root = api_dir.parent.parent.parent
        default_output = root / SAMPLE_DIRECTORY / OUTPUT_NAME
        return absolute, default_output, root

    raise TemplateNotFoundError(
        "template not found {0!r}; checked {1}".format(
            INPUT_NAME, ", ".join(checked)
        )
    )


def write_file(path, data, mode=PDF_FILE_MODE):
    # type: (Path, bytes, int) -> None
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(data)
    os.chmod(path, mode)


def render_document(input_path):
    # type: (Path) -> bytes
    html = input_path.read_bytes()
    base_url = input_path.resolve().parent.as_uri() + "/"
    document = Document(
        pages=[Page(source=Content.from_html(html, base=base_url))],
        page_size="A4",
        margin=Margin(top=0.0, right=0.0, bottom=0.0, left=0.0),
        background=True,
        smart_shrinking=False,
        allow_local_files=True,
    )
    return document.pdf()


def run(argv=None):
    # type: (list | None) -> int
    parser = argparse.ArgumentParser(
        prog="generate",
        description="Render testdata/golden/python_api/architecture-diagram.html",
    )
    parser.add_argument(
        "--output",
        default="",
        help="PDF output path (default: output/python/architecture-diagram.pdf)",
    )
    args = parser.parse_args(argv)

    input_path, default_output, _repo_root = resolve_template_paths()
    output_path = Path(args.output) if args.output else default_output
    if not output_path.is_absolute():
        output_path = (Path.cwd() / output_path).resolve()

    pdf = render_document(input_path)
    got = page_count(pdf)
    if got != WANT_PAGES:
        raise UnexpectedPageCountError(
            "unexpected page count: {0} pages = {1}, want {2}".format(
                input_path, got, WANT_PAGES
            )
        )

    write_file(output_path, pdf)
    print(
        "generated {0} ({1} pages, {2} bytes)".format(
            output_path, WANT_PAGES, len(pdf)
        )
    )
    return 0


def main():
    # type: () -> None
    try:
        raise SystemExit(run())
    except Exception as exc:  # noqa: BLE001 - CLI boundary
        print("generate:", exc, file=sys.stderr)
        raise SystemExit(1)


if __name__ == "__main__":
    main()
